packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -1432,38 +1363,136 @@ class TextPainter {
-    if (position.offset < 0) {
-      // TODO(LongCatIsLooong): make this case impossible; see https://github.com/flutter/flutter/issues/79495
-      return null;
-    }
-    return switch (_computeCaretMetrics(position)) {
-      _LineCaretMetrics(:final double fullHeight) => fullHeight,
-      _EmptyLineCaretMetrics() => null,
-    };
+    final TextBox textBox = _getOrCreateLayoutTemplate().getBoxesForRange(0, 1, boxHeightStyle: ui.BoxHeightStyle.strut).single;
+    return textBox.toRect().height;
   }
-
+  bool _isNewlineAtOffset(int offset) => 0 <= offset && offset < plainText.length
+                                      && WordBoundary._isNewline(plainText.codeUnitAt(offset));
   // Cached caret metrics. This allows multiple invokes of [getOffsetForCaret] and
   // [getFullHeightForCaret] in a row without performing redundant and expensive
   // get rect calls to the paragraph.
-  late _CaretMetrics _caretMetrics;
+  //
+  // The cache implementation assumes there's only one cursor at any given time.
+  late _LineCaretMetrics _caretMetrics;
 
-  // Checks if the [position] and [caretPrototype] have changed from the cached
-  // version and recomputes the metrics required to position the caret.
-  _CaretMetrics _computeCaretMetrics(TextPosition position) {
+  // This function returns the caret's offset and height for the given
+  // `position` in the text, or null if the paragraph is empty.
+  //
+  // For a TextPosition, typically when its TextAffinity is downstream, the
+  // corresponding I-beam caret is anchored to the leading edge of the
+  // `offset`-th character in the text. When the TextAffinity is upstream, the
+  // I-beam is then anchored to the trailing edge of the preceding character,
+  // except for a few edge cases:
+  //
+  // 1. empty paragraph: this method returns null and the caller handles this
+  //    case.
+  //
+  // 2. (textLength, downstream), the end-of-text caret when the text is not
+  //    empty: it's placed next to the trailing edge of the last line of the
+  //    text, in case the text and its last bidi run have different writing
+  //    directions. See the `_computeEndOfTextCaretAnchorOffset` method for more
+  //    details.
+  //
+  // 3. (0, upstream), which isn't a valid position, but it's not a conventional
+  //    "invalid" caret location either (the offset isn't negative). For
+  //    historical reasons, this is treated as (0, downstream).
+  //
+  // 4. (x, upstream) where x - 1 points to a line break character. The caret
+  //    should be displayed at the beginning of the newline instead of at the
+  //    end of the previous line. Converts the location to (x, downstream). The
+  //    choice we makes in 5. allows us to still check (x - 1) in case x points
+  //    to a multi-code-unit character.
+  //
+  // 5. (x, downstream || upstream), where x points to a multi-code-unit
+  //    character. There's no perfect caret placement in this case. Here we chose
```

Comment:
Meaning x could be just before the start of the character or it could be inside of it?

Reply:
`x ∈ [grapheme_start, grapheme_end)` where `grapheme_end > grapheme_start + 1`